lib/transport-common.js

```diff
@@ -0,0 +1,105 @@
+//            if not null must contain 'application'
+//            (if it doesn't default application will be used)
+//            can contain 'connectPolicy' if there is no
+//            connectPolicy default SimpleConnectPolicy will be used
+//            if it contains heartbeatInterval then heartbeat on JSTP
+//            connection will be started with it
+//   options - will be destructured and passed directly to connFactory
+//             The last argument of options is optional callback
+//             that will be called when connection is established
+//
+const createConnectionFactory = (
+  connFactory, transportFactory
+) => (appName, client, ...options) => {
+  const callback = common.extractCallback(options);
+  connFactory(...options, (error, rawConnection) => {
+    if (error) return callback(error);
+
+    const transport = transportFactory(rawConnection);
+    if (!client) {
+      client = { application: new Application('jstp', {}) };
+    } else if (!client.application) {
+      client.application = new Application('jstp', {});
+    }
+    if (!client.connectPolicy) {
+      client.connectPolicy = new SimpleConnectPolicy();
+    }
+    const connection = new Connection(transport, null, client);
+    client.connectPolicy.connect(appName, connection, (error, connection) => {
+      if (error) return callback(error, connection);
+      if (client.heartbeatInterval) {
+        connection.startHeartbeat(client.heartbeatInterval);
+      }
+      callback(null, connection);
+    });
+  });
+};
+
+// Same as createConnectionFactory but will also perform inspect of specified
+// interfaces.
+//   interfaces - array of interface names to perform inspect on
+//
+const createConnectionAndInspectFactory = (connFactory, transportFactory) => {
+  const connect = createConnectionFactory(connFactory, transportFactory);
+  return (appName, client, interfaces, ...options) => {
+    const callback = common.extractCallback(options);
+    connect(appName, client, ...options, (error, connection) => {
+      if (error) return callback(error);
+
+      Promise.all(interfaces.map(name => new Promise((resolve, reject) => {
+        connection.inspectInterface(name, (error, proxy) => {
+          if (error) {
+            reject(error);
+          } else {
+            resolve(proxy);
+          }
+        });
+      }))).then((proxies) => {
+        const api = proxies.reduce((acc, proxy, idx) => {
+          const name = interfaces[idx];
+          acc[name] = proxy;
+          return acc;
+        }, Object.create(null));
+        callback(null, connection, api);
+      }).catch((error) => {
+        callback(error, connection);
+      });
+    });
+  };
+};
+
+// Utility method to create function to produce servers with
+// serverFactory.
+// If options is an array then wraps it as { applications: options }.
+//
+const createServerFactory = serverFactory => (options, ...other) => {
```

Comment:
This function should take a class to avoid creating unnecessary closures in concrete server implementations.